tiledb/cloud/taskgraphs/_codec.py

```diff
@@ -71,3 +130,102 @@ def _replace_sentinel(
     "native": cloudpickle.loads,
     "python_pickle": cloudpickle.loads,
 }
+
+
+@attrs.define(frozen=True, slots=False)
+class BinaryResult(TDBJSONEncodable):
+    """Container for a binary-encoded value, decoded on-demand.
+
+    This is used to store results obtained from the server, such that it is not
+    necessary to decode them between stages, and they only are decoded upon
+    request.
+    """
+
+    format: str
+    """The format of the data."""
+    data: bytes
+    """The binary data itself."""
+
+    @classmethod
+    def from_response(cls, resp: urllib3.HTTPResponse) -> "BinaryResult":
+        """Reads a urllib3 response into an encoded result."""
+        full_mime = resp.getheader("Content-type", "application/octet-stream")
+        mime, _, _ = full_mime.partition(";")
+        mime = mime.strip()
+        format = _MIME_TO_FORMAT.get(mime, "mime:" + mime)
+        data = resp.data
+        return cls(format, data)
+
+    def decode(self) -> types.NativeValue:
+        """Decodes this result into native Python data."""
+        # This is not lock-protected because we're ok with decoding twice.
```

Comment:
is there a hidden performance problem here? if it someday became the case that we were decoding twice (or more), would we know?